### PR TITLE
common: introduce stable logger /w examples in DNS 

### PR DESCRIPTION
**Pull request by snowp**
Adds a new logging macro that allows for logging with a stable identifier. This is intended to allow
a log sink to be able to key off the identifier, simplifying the process of having a log sink react to specific
log messages (e.g. to emit specific events when certain log lines are hit). 

The main alternative to this is using some kind of log parsing to understand when a certain log is hit, which ends up being brittle as code is refactored. Using a stable identifier we can ensure via code review that the stable id remains as log lines are moved around, refactored, etc.

This PR includes applying the new log macro to the c-ares DNS resolver to demonstrate it's usage.

Signed-off-by: Snow Pettersen <snowp@lyft.com>

Risk Level: Medium
Testing: Manual testing, will add coverage
Docs Changes: n/a
Release Notes: n/a (will add)
Platform Specific Features: n/a

### CODE REVIEW

**Comment by snowp** [comment]
Note that this does not support log levels as it bypasses spdlog. spdlog does not support any form of structured logging, so going through spdlog would mean either 1) extracting out the stable name from the formatted log string (not much better than having a log sink itself try to parse out information) or 2) by having different spdlog loggers set up for each call site (which seemed very heavy handed).

It is also always on as written, even when no log sink is consuming this data. The consequence of this is mostly that the fmt::format is evaluated. I imagine we want to avoid this, but figured I'd put up the PR first before investing the time into figuring out a good solution there.

**Comment by snowp** [comment]
An alternative API might be something like

`ENVOY_LOG_EVENT(name, level, msg)` which delegates to `ENVOY_LOG(level, msg)` while also emitting an `ENVOY_STABLE_LOG(name, msg)`

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this looks like it's on the right track. Flushing out some initial comments.

/wait

File: source/common/network/dns_impl.cc [hunk 1]
```diff
@@ -253,6 +260,8 @@ void DnsResolverImpl::onAresSocketStateChange(os_fd_t fd, int read, int write) {
 
 ActiveDnsQuery* DnsResolverImpl::resolve(const std::string& dns_name,
                                          DnsLookupFamily dns_lookup_family, ResolveCb callback) {
+  ENVOY_LOG_EVENT(debug, "cares_dns_resolution_started", "dns resolution for {} started", dns_name);
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc:263]: inconsistent use of "cares" in this event name vs the other names.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this looks right to me. @envoyproxy/maintainers thoughts on this? This feels like something we can ship and iterate on.

/wait

**Comment by ggreenway** [comment]
What's an example output from one of these logs? I'm not sure I 100% understand the goal, but it seems like not having log-levels will quickly become troublesome.

**Comment by snowp** [comment]
@ggreenway As written there is no new sink handling built in for these logs - the logs are currently sent both to `ENVOY_LOG(level, ...)` as well as the new `logWithStableName` function on the sink (which is a noop by default). 

What we're trying to do in EM is allow hooking up a custom sink that can intercept these log events and emit analytic events, allowing us to incorporate Envoy events into debugging. This would allow us to correlate sequences of events where we see requests fail with Envoy events, e.g. failing connection setup, DNS timing out etc. 

This could also be built up into a more sophisticated event tracing library where we can associated events with each other, so that more sophisticated tooling can be built to analyze or render entire stream lifetimes, but for now this simple "log the event with a stable name" would be a good start for the EM use case.

**Comment by ggreenway** [comment]
Got it; that makes sense. SGTM, but please put a big doc comment above the new top-level macro so people know what it is for and when to use it.

**Comment by snowp** [comment]
I updated this to check the log level for these new events as well, which ended up being pretty simple since can just check the log level of the associated component.

**Review by mattklein123 — APPROVED** [review]
LGTM, thanks

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**